src/fat_beagle.cpp

```diff
@@ -506,57 +543,87 @@ PhyloGradient FatBeagle::Gradient(const UnrootedTree &in_tree) const {
                                   phylo_model_->GetSiteModel()->GetRateGradient());
     auto [log_likelihood, unscaled_category_gradient] =
         BranchGradientInternals(tree.Topology(), tree.BranchLengths(), dQ);
-    gradient["site_model"] =
+    gradient[PhyloGradient::site_model_key_] =
         DiscreteSiteModelGradient(tree.BranchLengths(), unscaled_category_gradient);
   }
 
   // We want the fixed node to have a zero gradient.
   branch_length_gradient[tree.Topology()->Children()[1]->Id()] = 0.;
-  gradient["branch_lengths"] = branch_length_gradient;
+  gradient[PhyloGradient::branch_lengths_key_] = branch_length_gradient;
 
   return {log_likelihood, gradient};
 }
 
-PhyloGradient FatBeagle::Gradient(const RootedTree &tree) const {
+PhyloGradient FatBeagle::Gradient(
+    const RootedTree &tree, std::optional<PhyloFlags> flags /*= std::nullopt*/) const {
   // Scale time with clock rate.
   std::vector<double> branch_lengths = tree.BranchLengths();
   const std::vector<double> &rates = tree.GetRates();
   for (size_t i = 0; i < tree.BranchLengths().size() - 1; i++) {
     branch_lengths[i] *= rates[i];
   }
-
   // Calculate branch length gradient and log likelihood.
   EigenMatrixXd dQ =
       BuildDifferentialMatrices(*phylo_model_->GetSubstitutionModel(),
                                 phylo_model_->GetSiteModel()->GetCategoryRates());
   auto [log_likelihood, branch_gradient] =
       BranchGradientInternals(tree.Topology(), branch_lengths, dQ);
-
   GradientMap gradient;
-
-  gradient["branch_lengths"] = branch_gradient;
-  // Calculate substitution model parameter gradient, if needed.
-  if (phylo_model_->GetSubstitutionModel()->GetRates().size() > 0) {
-    FatBeagle *mutable_this = const_cast<FatBeagle *>(this);
-    gradient["substitution_model"] = SubstitutionModelGradient<RootedTree>(
-        FatBeagle::StaticRootedLogLikelihood, mutable_this, tree);
+  gradient[PhyloGradient::branch_lengths_key_] = branch_gradient;
+  // Calculate Substitution Model Gradients (Rates and/or Frequencies), if flagged for.
+  bool do_subst_grad = PhyloFlags::IsFlagChecked(
+      flags, PhyloFlags::gradient_options_.substitution_model_);
+  bool do_rates_grad = PhyloFlags::IsFlagChecked(
+      flags, PhyloFlags::gradient_options_.substitution_model_rates_);
+  bool do_freqs_grad = PhyloFlags::IsFlagChecked(
+      flags, PhyloFlags::gradient_options_.substitution_model_frequencies_);
+  if (do_subst_grad || do_rates_grad || do_freqs_grad) {
+    if (phylo_model_->GetSubstitutionModel()->GetRates().size() > 0) {
+      FatBeagle *mutable_this = const_cast<FatBeagle *>(this);
```

Comment:
@ognian- did you discuss this with Dave?

Reply:
The cast following this line was no longer necessary since these functions are not overloaded anymore.

Reply:
I'm asking about casting away const, which isn't best practice and still appears in the code.

Reply:
Okay, gotcha.  Just FYI, I didn't write that, it says it is from PR #330.

Reply:
Since we have only implemented JC69, HKY and GTR we have either no parameters or both rate and frequency parameters. If one day we implement F81 (or a GY94 codon model) the derivatives for the frequency parameters won't be computed because of line 581.

Reply:
Perhaps @4ment can remind us why we have to cast away const here, and insert a comment so that I don't ask again?

Reply:
I think it is because `FatBeagle::Gradient` calls (indirectly through `FatBeagle::SubstitutionModelGradient`) `FatBeagle::StaticUnrootedLogLikelihood` or `FatBeagle::StaticRootedLogLikelihood` and both of them require a non constant FatBeagle reference.

Reply:
These eventually call LogLikelihood, which is const, which makes me wonder if we could just put some consts in the static call arguments and get rid of the cast. @davidrich27 if have time I'd like to see if this is possible. Thx!

Reply:
@matsen Static function calls now take const pointers.